File: cpp/45.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        if(nums.size()<=1) return 0;
        int maxReach = nums[0];
        int curMaxReach = nums[0];
        int step = 1;
        for(int i=1; i<=min((int)nums.size(),maxReach); i++){
            curMaxReach = max(curMaxReach, i+nums[i]);
            if(i==nums.size()-1) break;
            if(i==maxReach){
                maxReach = curMaxReach;
                step++;
            }
        }
        return step;
    }
};

File: golang/45/45.go
package main

func jump(nums []int) int {
	end := 0
	maxPos := 0
	step := 0
	for i := 0; i < len(nums)-1; i++ {
		maxPos = max(maxPos, nums[i]+i)
		if i == end {
			end = maxPos
			step++
		}
	}
	return step
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
